spark/src/main/scala/org/apache/comet/rules/CometScanRule.scala

```diff
@@ -105,8 +105,49 @@ case class CometScanRule(session: SparkSession) extends Rule[SparkPlan] {
           return withInfos(scanExec, fallbackReasons.toSet)
         }
 
-        val scanImpl = COMET_NATIVE_SCAN_IMPL.get()
-        if (scanImpl == CometConf.SCAN_NATIVE_DATAFUSION && !COMET_EXEC_ENABLED.get()) {
+        var scanImpl = COMET_NATIVE_SCAN_IMPL.get()
+
+        // if scan is auto then pick best available scan
+        if (scanImpl == SCAN_AUTO) {
+          // TODO these checks are not yet exhaustive. For example, native_datafusion does
+          //  not support reading from object stores such as S3 yet
+
+          val typeChecker = CometScanTypeChecker(SCAN_NATIVE_ICEBERG_COMPAT)
+          val schemaSupported =
+            typeChecker.isSchemaSupported(scanExec.requiredSchema, fallbackReasons)
+          val partitionSchemaSupported =
+            typeChecker.isSchemaSupported(r.partitionSchema, fallbackReasons)
+
+          // additional checks for known issues
+          def isComplexType(dt: DataType): Boolean = dt match {
+            case _: StructType | _: ArrayType | _: MapType => true
+            case _ => false
+          }
+
+          def hasKnownIssues(dataType: DataType): Boolean = {
+            dataType match {
+              case s: StructType => s.exists(field => hasKnownIssues(field.dataType))
+              case a: ArrayType => hasKnownIssues(a.elementType)
+              case m: MapType => isComplexType(m.keyType) || isComplexType(m.valueType)
+              case _ => false
```

Comment:
This seems to indicate that if we have an array type with the element type being a struct with primitive fields then there are no known issues. But if we have a map with a value type of the same struct with primitive fields then it is not supported?

Reply:
I made these checks quite conservative for now. Hopefully we can relax the rules at some point.

Reply:
Right. 
For `Array(elementType = Struct(PrimitiveType)`, `hasKnownIssues` would return `false` 
For `Map(..., valueType=Struct(PrimitiveType)`,  `hasKnownIssues` would return `true`
Is that what is intended? Not sure why that would be the case, tbh.

Reply:
The current implementation of `hasKnownIssues` only returns `true` for maps of complex types (whether at the top level of within other complex types), because at the time of writing, there was a known issue with maps of structs. I'm not sure if this is still the case.

Reply:
I see. Thank you for the explanation :)